File: Python/count-substrings-with-k-frequency-characters-i.py
# Time:  O(n + 26)
# Space: O(26)

# freq table, two pointers, sliding window
class Solution(object):
    def numberOfSubstrings(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        def count():
            cnt = [0]*26
            result= left = 0
            for right in xrange(len(s)):
                cnt[ord(s[right])-ord('a')] += 1
                while cnt[ord(s[right])-ord('a')] == k:
                    cnt[ord(s[left])-ord('a')] -= 1
                    left += 1
                result += right-left+1
            return result
                
        return (len(s)+1)*len(s)//2-count()

File: C++/count-substrings-with-k-frequency-characters-i.cpp
// Time:  O(n + 26)
// Space: O(26)

// freq table, two pointers, sliding window
class Solution {
public:
    int numberOfSubstrings(string s, int k) {
        const auto& count = [&]() {
            int result = 0;
            vector<int> cnt(26);
            for (int right = 0, left = 0; right < size(s); ++right) {
                ++cnt[s[right] - 'a'];
                for (; cnt[s[right] - 'a'] == k; ++left) {
                    --cnt[s[left] - 'a'];
                }
                result += right - left + 1;
            }
            return result;
        };

        return (size(s) + 1) * size(s) / 2 - count();
    }
};
